### PR TITLE
#71 get candidates slots

### PR DESCRIPTION
**Pull request by alesyalis**
create GetNewCandidatesSlots for Recruiters return only candidates who  have type new 
create  GetUnAssignedCandidatesSlots return only candidates who not assigned, and have type of interview - recruiter, tech1, tech2

### CODE REVIEW

File: Recrutify/Recrutify.DataAccess/Models/AssignedCandidate.cs [hunk 1]
```diff
@@ -11,14 +11,10 @@ public class AssignedCandidate
...
-        public Status Status { get; set; }
-
-        public Guid ProjectId { get; set; }
-
         public string Email { get; set; }
 
         public string Skype { get; set; }
 
-        public CandidatePrimarySkill PrimarySkill { get; set; }
+        public IEnumerable<ProjectResultAssigned> ProjectResults { get; set; }
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Models/AssignedCandidate.cs]: not list

File: Recrutify/Recrutify.Services/DTOs/AssignedCandidateDTO.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        public Guid Id { get; set; }
+
+        public string Name { get; set; }
+
+        public IEnumerable<int> BestTimeToConnect { get; set; }
+
+        public string Email { get; set; }
+        
+        public IEnumerable<ProjectResultAssignedDTO> ProjectResults { get; set; }
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/DTOs/AssignedCandidateDTO.cs:16]: not list

Comment (alesyalis) [reply Recrutify/Recrutify.Services/DTOs/AssignedCandidateDTO.cs:16]: done

File: Recrutify/Recrutify.DataAccess/Models/ProjectResultAssigned.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+
+namespace Recrutify.DataAccess.Models
+{
+    public class ProjectResultAssigned
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Models/ProjectResultAssigned.cs:5]: AssignedCandidateProjectResult, 
also use inheritance with ProjectResult, not to duplicate

File: Recrutify/Recrutify.Host/Configuration/OData/CandidatesConfiguration.cs [hunk 4]
```diff
@@ -9,11 +9,12 @@ public class CandidatesConfiguration : IModelConfiguration
         public void Apply(ODataModelBuilder builder, ApiVersion apiVersion, string routePrefix)
         {
             builder.EntitySet<CandidateDTO>("Candidates");
+            builder.EntityType<CandidateDTO>();
             builder.ComplexType<ProjectResultDTO>();
             builder.ComplexType<CandidatePrimarySkillDTO>();
             builder.EntityType<CandidateDTO>().Collection
-                   .Function("GetByProject")
-                   .ReturnsCollectionFromEntitySet<CandidateDTO>("Candidates");
+                    .Function("GetByProject")
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Host/Configuration/OData/CandidatesConfiguration.cs:16]: extra spaces

File: Recrutify/Recrutify.Host/Configuration/OData/ScheduleConfiguration.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    public class ScheduleConfiguration : IModelConfiguration
+    {
+        public void Apply(ODataModelBuilder builder, ApiVersion apiVersion, string routePrefix)
+        {
+            builder.EntitySet<AssignedCandidateDTO>("Schedule");
+            builder.EntityType<AssignedCandidateDTO>().Collection
+                  .Function("GetNewCandidatesSlots")
+                  .ReturnsCollectionFromEntitySet<AssignedCandidateDTO>("Schedule");
+            builder.EntityType<AssignedCandidateDTO>().Collection
+                  .Function("GetUnAssignedCandidatesSlots")
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Host/Configuration/OData/ScheduleConfiguration.cs:16]: Correct spelling Unassigned

File: Recrutify/Recrutify.DataAccess/Models/ProjectResultAssigned.cs (hunk 3 above)
Comment (alesyalis) [reply Recrutify/Recrutify.DataAccess/Models/ProjectResultAssigned.cs]: done

File: Recrutify/Recrutify.Host/Configuration/OData/CandidatesConfiguration.cs (hunk 4 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Host/Configuration/OData/CandidatesConfiguration.cs]: done

File: Recrutify/Recrutify.Host/Configuration/OData/ScheduleConfiguration.cs (hunk 5 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Host/Configuration/OData/ScheduleConfiguration.cs]: done

File: Recrutify/Recrutify.DataAccess/Models/AssignedCandidate.cs (hunk 1 above)
Comment (alesyalis) [reply Recrutify/Recrutify.DataAccess/Models/AssignedCandidate.cs]: done

File: Recrutify/Recrutify.DataAccess/Models/BaseProjectResult.cs [hunk 6]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Recrutify.DataAccess.Models
+{
+    public class BaseProjectResult
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Models/BaseProjectResult.cs:9]: abstract

File: Recrutify/Recrutify.Services/Services/ProjectService.cs [hunk 7]
```diff
@@ -56,7 +56,7 @@ public async Task<List<ProjectDTO>> GetAllAsync()
 
         public IQueryable<ShortProjectDTO> GetShort()
         {
-            var projects = _projectRepository.Get();
+            var projects = _projectRepository.GetShortProject();
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/ProjectService.cs:59]: GetShort

File: Recrutify/Recrutify.DataAccess/Models/ProjectResultAssigned.cs [hunk 8]
```diff
@@ -0,0 +1,8 @@
+﻿using System;
+
+namespace Recrutify.DataAccess.Models
+{
+    public class ProjectResultAssigned : BaseProjectResult
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Models/ProjectResultAssigned.cs:5]: rename

File: Recrutify/Recrutify.DataAccess/Models/AssignedCandidate.cs [hunk 9]
```diff
@@ -11,14 +11,10 @@ public class AssignedCandidate
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Models/AssignedCandidate.cs:11]: rename class

File: Recrutify/Recrutify.DataAccess/Models/ProjectResultAssigned.cs (hunk 8 above)
Comment (alesyalis) [reply Recrutify/Recrutify.DataAccess/Models/ProjectResultAssigned.cs]: done

File: Recrutify/Recrutify.DataAccess/Models/AssignedCandidate.cs (hunk 9 above)
Comment (alesyalis) [reply Recrutify/Recrutify.DataAccess/Models/AssignedCandidate.cs:11]: done

File: Recrutify/Recrutify.Services/Services/ProjectService.cs (hunk 7 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Services/ProjectService.cs]: done

File: Recrutify/Recrutify.DataAccess/Models/BaseProjectResult.cs (hunk 6 above)
Comment (alesyalis) [reply Recrutify/Recrutify.DataAccess/Models/BaseProjectResult.cs]: done

File: Recrutify/Recrutify.DataAccess/Models/ProjectResultScheduleCandidate.cs [hunk 10]
```diff
@@ -0,0 +1,8 @@
+﻿using System;
+
+namespace Recrutify.DataAccess.Models
+{
+    public class ProjectResultScheduleCandidate : BaseProjectResult
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Models/ProjectResultScheduleCandidate.cs:5]: ScheduleProjectResultInfo

File: Recrutify/Recrutify.Host/Configuration/Profiles/CandidateProfile.cs [hunk 11]
```diff
@@ -1,4 +1,5 @@
 ﻿using System;
+using System.Linq;
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Host/Configuration/Profiles/CandidateProfile.cs:2]: remove

File: Recrutify/Recrutify.Services/DTOs/AssignedCandidateDTO.cs [hunk 12]
```diff
@@ -0,0 +1,18 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace Recrutify.Services.DTOs
+{
+    public class AssignedCandidateDTO
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/DTOs/AssignedCandidateDTO.cs:6]: rename

File: Recrutify/Recrutify.Services/DTOs/ProjectResultAssignedDTO.cs [hunk 13]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+
+namespace Recrutify.Services.DTOs
+{
+    public class ProjectResultAssignedDTO
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/DTOs/ProjectResultAssignedDTO.cs:5]: rename

File: Recrutify/Recrutify.Host/Configuration/OData/CandidatesConfiguration.cs [hunk 14]
```diff
@@ -9,6 +9,7 @@ public class CandidatesConfiguration : IModelConfiguration
         public void Apply(ODataModelBuilder builder, ApiVersion apiVersion, string routePrefix)
         {
             builder.EntitySet<CandidateDTO>("Candidates");
+            builder.EntityType<CandidateDTO>();
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Host/Configuration/OData/CandidatesConfiguration.cs:12]: is it necessary?

File: Recrutify/Recrutify.Services/DTOs/ProjectResultAssignedDTO.cs (hunk 13 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/DTOs/ProjectResultAssignedDTO.cs:5]: done

File: Recrutify/Recrutify.Services/DTOs/AssignedCandidateDTO.cs (hunk 12 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/DTOs/AssignedCandidateDTO.cs]: done

File: Recrutify/Recrutify.Host/Configuration/Profiles/CandidateProfile.cs (hunk 11 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Host/Configuration/Profiles/CandidateProfile.cs]: done

File: Recrutify/Recrutify.Services/Services/CandidateService.cs [hunk 15]
```diff
@@ -44,6 +44,32 @@ public IQueryable<CandidateDTO> GetByProject(Guid projectId)
...
+            foreach (var r in result)
+            {
+                var candidat = candidates.FirstOrDefault(x => x.Id == r.Id);
+                r.ProjectResult = _mapper.Map<ScheduleCandidateProjectResultDTO>(candidat.ProjectResults.FirstOrDefault(x => x.ProjectId == projectId));
+            }
+
+            return result;
+        }
+
+        public IEnumerable<ScheduleCandidateInfoDTO> GetUnassignedCandidateByProject(Guid projectId)
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/CandidateService.cs:60]: what is being done in order to return **only unassigned** users?

Comment (elizavetapavliv) [reply Recrutify/Recrutify.Services/Services/CandidateService.cs:60]: why Candidat**e** if its list?

File: Recrutify/Recrutify.Host/Configuration/OData/CandidatesConfiguration.cs (hunk 14 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Host/Configuration/OData/CandidatesConfiguration.cs]: done

File: Recrutify/Recrutify.Services/Services/Abstract/ICandidateService.cs [hunk 16]
```diff
@@ -27,5 +27,9 @@ public interface ICandidateService
         Task BulkCreateTestFeedbacksAsync(BulkCreateTestFeedbackDTO bulkCreateTestFeedbackDTO, Guid projectId);
 
         Task BulkUpdateStatusReasonAsync(BulkUpdateStatusDTO bulkUpdateStatusDTO, Guid projectId);
+
+        IEnumerable<ScheduleCandidateInfoDTO> GetNewCandidateByProject(Guid projectId);
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/Abstract/ICandidateService.cs:31]: rename methods like in controller

File: Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs [hunk 17]
```diff
@@ -23,6 +23,16 @@ public IQueryable<Candidate> GetByProject(Guid projectId)
             return GetCollection().AsQueryable().Where(x => x.ProjectResults.Any(y => y.ProjectId == projectId));
         }
 
+        public IQueryable<Candidate> GetNewCandidateByProject(Guid projectId)
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs:26]: GetNewCandidate**s**ByProject

File: Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs [hunk 18]
```diff
@@ -23,6 +23,16 @@ public IQueryable<Candidate> GetByProject(Guid projectId)
...
+        {
+            return GetCollection().AsQueryable().Where(x => x.ProjectResults.Any(y => y.ProjectId == projectId && y.Status == Status.New));
+        }
+
+        public IQueryable<Candidate> GetUnassignedCandidateByProject(Guid projectId)
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs:31]: GetUnassignedCandidate**s**ByProject

File: Recrutify/Recrutify.Services/Services/Abstract/ICandidateService.cs (hunk 16 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Services/Abstract/ICandidateService.cs]: done

File: Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs (hunk 17 above)
Comment (alesyalis) [reply Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs]: done

File: Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs (hunk 18 above)
Comment (alesyalis) [reply Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs]: done

File: Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs [hunk 19]
```diff
@@ -23,6 +23,16 @@ public IQueryable<Candidate> GetByProject(Guid projectId)
...
         }
 
+        public IQueryable<Candidate> GetCandidatesPassedTestByProject(Guid projectId)
+        {
+            return GetCollection().AsQueryable().Where(x => x.ProjectResults.Any(y => y.ProjectId == projectId && y.Status == Status.Test));
+        }
+
+        public IQueryable<Candidate> GetUnassignedCandidatesByProject(Guid projectId)
+        {
+            return GetCollection().AsQueryable().Where(x => x.ProjectResults.Any(p => p.ProjectId == projectId && !p.IsAssigned && (p.Status == Status.RecruiterInterview || p.Status == Status.TechInterviewOneStep || p.Status == Status.TechInterviewSecondStep)));
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs:33]: rename IsAssigned to IsAssignedOnInterview, and in dto too

File: Recrutify/Recrutify.Services/Services/CandidateService.cs [hunk 20]
```diff
@@ -44,6 +44,32 @@ public IQueryable<CandidateDTO> GetByProject(Guid projectId)
             return _mapper.ProjectTo<CandidateDTO>(candidates);
         }
 
+        public IEnumerable<ScheduleCandidateInfoDTO> GetCandidatesPassedTestSlots(Guid projectId)
+        {
+            var candidate = _candidateRepository.GetCandidatesPassedTestByProject(projectId);
+            var candidatesDtos = _mapper.ProjectTo<ScheduleCandidateInfoDTO>(candidate).ToList();
+            foreach (var c in candidatesDtos)
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/CandidateService.cs:51]: x rename to c
c to candidateDto

File: Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs (hunk 19 above)
Comment (elizavetapavliv) [reply Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs:33]: + scripts

File: Recrutify/Recrutify.Services/Services/CandidateService.cs (hunk 20 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Services/CandidateService.cs]: done

File: Recrutify/Recrutify.Services/Services/CandidateService.cs [hunk 21]
```diff
@@ -44,6 +44,32 @@ public IQueryable<CandidateDTO> GetByProject(Guid projectId)
...
+        }
+
+        public IEnumerable<ScheduleCandidateInfoDTO> GetUnassignedCandidatesSlots(Guid projectId)
+        {
+            var candidates = _candidateRepository.GetUnassignedCandidatesByProject(projectId);
+            var candidatesDtos = _mapper.ProjectTo<ScheduleCandidateInfoDTO>(candidates).ToList();
+            foreach (var candidateDto in candidatesDtos)
+            {
+                var candidate = candidates.FirstOrDefault(c => c.Id == candidateDto.Id);
+                candidateDto.ProjectResult = _mapper.Map<ScheduleCandidateProjectResultDTO>(candidate.ProjectResults.FirstOrDefault(x => x.ProjectId == projectId));
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/CandidateService.cs:67]: c

Comment (elizavetapavliv) [reply Recrutify/Recrutify.Services/Services/CandidateService.cs:67]: p

File: Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs (hunk 19 above)
Comment (alesyalis) [reply Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs:33]: done

**Review by elizavetapavliv — APPROVED** [review]

**Pull request merged**